File: XRViewer/AppStateController/AppState.m
#import "AppState.h"

@implementation AppState

- (instancetype)copyWithZone:(NSZone *)zone
{
    AppState *copy = [AppState new];
    [copy setShowOptions:[self showOptions]];
    [copy setShowMode:[self showMode]];
    [copy setWebXR:[self webXR]];
    [copy setARRequest:[self aRRequest]];
    [copy setComputerVisionFrameRequested:[self computerVisionFrameRequested]];
    [copy setShouldRemoveAnchorsOnNextARSession: [self shouldRemoveAnchorsOnNextARSession]];
    [copy setSendComputerVisionData:[self sendComputerVisionData]];
    [copy setShouldShowLiteModePopup:[self shouldShowLiteModePopup]];
    [copy setShouldShowSessionStartedPopup: [self shouldShowSessionStartedPopup]];
    [copy setNumberOfTimesSendNativeTimeWasCalled: [self numberOfTimesSendNativeTimeWasCalled]];
    [copy setUserGrantedSendingComputerVisionData: [self userGrantedSendingComputerVisionData]];
    [copy setAskedComputerVisionData: [self askedComputerVisionData]];
    [copy setUserGrantedSendingWorldStateData: [self userGrantedSendingWorldStateData]];
    [copy setAskedWorldStateData: [self askedWorldStateData]];

    return copy;
}

- (BOOL)isEqual:(id)theObject
{
    if (theObject == self)
    {
        return YES;
    }
    
    if ([self class] != [theObject class])
    {
        return NO;
    }
    
    if ([self showOptions] != [theObject showOptions])
    {
        return NO;
    }
    
    if ([self showMode] != [theObject showMode])
    {
        return NO;
    }
    
    if ([self webXR] != [theObject webXR])
    {
        return NO;
    }
    
    if ([self aRRequest] != [theObject aRRequest] && [[self aRRequest] isEqualToDictionary:[theObject aRRequest]] == NO)
    {
        return NO;
    }
    
    if ([self trackingState] != [theObject trackingState] && [[self trackingState] isEqualToString:[theObject trackingState]] == NO)
    {
        return NO;
    }

    if ([self computerVisionFrameRequested] != [theObject computerVisionFrameRequested])
    {
        return NO;
    }

    if ([self shouldRemoveAnchorsOnNextARSession] != [theObject shouldRemoveAnchorsOnNextARSession])
    {
        return NO;
    }

    if ([self sendComputerVisionData] != [theObject sendComputerVisionData])
    {
        return NO;
    }
    
    if ([self shouldShowLiteModePopup] != [theObject shouldShowLiteModePopup]) {
        return NO;
    }
    
    if ([self shouldShowSessionStartedPopup] != [theObject shouldShowSessionStartedPopup]) {
        return NO;
    }
    
    if ([self numberOfTimesSendNativeTimeWasCalled] != [theObject numberOfTimesSendNativeTimeWasCalled]) {
        return NO;
    }

    if ([self userGrantedSendingComputerVisionData] != [theObject userGrantedSendingComputerVisionData]) {
        return NO;
    }

    if ([self askedComputerVisionData] != [theObject askedComputerVisionData]) {
        return NO;
    }

    if ([self userGrantedSendingWorldStateData] != [theObject userGrantedSendingWorldStateData]) {
        return NO;
    }
    
    if ([self askedWorldStateData] != [theObject askedWorldStateData]) {
        return NO;
    }
    
    return YES;
}

- (NSUInteger)hash
{
    return [self showOptions] ^ [self showMode] ^ [self webXR] ^ [[self trackingState] hash] ^ [[self aRRequest] hash];
}

+ (instancetype)defaultState
{
    AppState *state = [AppState new];
    
    [state setShowMode:SHOW_MODE_BY_DEFAULT];
    [state setShowOptions:SHOW_OPTIONS_BY_DEFAULT];
    [state setShouldShowLiteModePopup:YES];
    [state setShouldShowSessionStartedPopup:POPUP_ENABLED_BY_DEFAULT];
    [state setNumberOfTimesSendNativeTimeWasCalled:0];
    [state setUserGrantedSendingComputerVisionData:USER_GRANTED_SENDING_COMPUTER_VISION_DATA_BY_DEFAULT];
    [state setUserGrantedSendingWorldStateData:SendWorldSensingDataAuthorizationStateDenied];
    [state setAskedComputerVisionData:NO];
    [state setAskedWorldStateData:NO];

    // trackingstate default is nil ?
    
    return state;
}

- (instancetype)updatedShowMode:(ShowMode)showMode
{
    [self setShowMode:showMode];
    return self;
}

- (instancetype)updatedShowOptions:(ShowOptions)showOptions
{
    [self setShowOptions:showOptions];
    return self;
}

- (instancetype)updatedWebXR:(BOOL)webXR
{
    [self setWebXR:webXR];
    return self;
}

- (instancetype)updatedWithARRequest:(NSDictionary *)dict
{
    [self setARRequest:dict];
    return self;
}

@end

File: XRViewer/AppStateController/AppState.swift
import Foundation

/**
 The app internal state
 */
@objc class AppState: NSObject, NSCopying {
    @objc var aRRequest: [AnyHashable : Any] = [:]
    var showMode: ShowMode = .nothing
    var showOptions: ShowOptions = .init(rawValue: 0)
    var webXR = false
    var geometryArrays = false
    var computerVisionFrameRequested = false
    var shouldRemoveAnchorsOnNextARSession = false
    var sendComputerVisionData = false
    var shouldShowSessionStartedPopup = false
    var shouldShowLiteModePopup = false
    var numberOfTimesSendNativeTimeWasCalled: Int = 0
    var numberOfTrackedImages: Int = 0
    @objc var userGrantedSendingComputerVisionData = false
    @objc var askedComputerVisionData = false
    @objc var userGrantedSendingWorldStateData: WebXRAuthorizationState = .notDetermined
    @objc var askedWorldStateData = false

    class func defaultState() -> AppState {
        let state = AppState()

        state.showMode = ShowMode.nothing
        state.showOptions = ShowOptions.init(rawValue: 0)
        state.shouldShowSessionStartedPopup = true
        state.geometryArrays = false
        state.shouldShowLiteModePopup = true
        state.numberOfTimesSendNativeTimeWasCalled = 0
        state.numberOfTrackedImages = 0
        state.userGrantedSendingComputerVisionData = false
        state.userGrantedSendingWorldStateData = .notDetermined
        state.askedComputerVisionData = false
        state.askedWorldStateData = false

        return state
    }

    func updatedShowMode(_ showMode: ShowMode) -> Self {
        self.showMode = showMode
        return self
    }

    func updatedShowOptions(_ showOptions: ShowOptions) -> Self {
        self.showOptions = showOptions
        return self
    }

    func updatedWebXR(_ webXR: Bool) -> Self {
        self.webXR = webXR
        return self
    }

    func updated(withARRequest dict: [AnyHashable : Any]?) -> Self {
        guard let dict = dict else { return self }
        aRRequest = dict
        return self
    }

    func copy(with zone: NSZone? = nil) -> Any {
        let copy = AppState()
        copy.showOptions = showOptions
        copy.showMode = showMode
        copy.webXR = webXR
        copy.geometryArrays = geometryArrays
        copy.aRRequest = aRRequest
        copy.computerVisionFrameRequested = computerVisionFrameRequested
        copy.shouldRemoveAnchorsOnNextARSession = shouldRemoveAnchorsOnNextARSession
        copy.sendComputerVisionData = sendComputerVisionData
        copy.shouldShowSessionStartedPopup = shouldShowSessionStartedPopup
        copy.shouldShowLiteModePopup = shouldShowLiteModePopup
        copy.numberOfTimesSendNativeTimeWasCalled = numberOfTimesSendNativeTimeWasCalled
        copy.numberOfTrackedImages = numberOfTrackedImages
        copy.userGrantedSendingComputerVisionData = userGrantedSendingComputerVisionData
        copy.askedComputerVisionData = askedComputerVisionData
        copy.userGrantedSendingWorldStateData = userGrantedSendingWorldStateData
        copy.askedWorldStateData = askedWorldStateData

        return copy
    }
}
